File: backend/src/main/kotlin/top/bearingwall/asya/service/AuditLogService.kt
package top.bearingwall.asya.service

import org.slf4j.LoggerFactory
import org.springframework.data.domain.Page
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Propagation
import org.springframework.transaction.annotation.Transactional
import top.bearingwall.asya.dto.AuditLogResponse
import top.bearingwall.asya.model.AuditActionType
import top.bearingwall.asya.model.AuditLog
import top.bearingwall.asya.repository.AuditLogRepository
import java.time.LocalDateTime
import java.time.ZoneOffset
import java.util.UUID

@Service
class AuditLogService(
    private val auditLogRepository: AuditLogRepository
) {
    private val log = LoggerFactory.getLogger(AuditLogService::class.java)

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    fun save(
        actorUuid: UUID?,
        actorName: String?,
        actorIp: String?,
        requestMethod: String? = null,
        requestPath: String? = null,
        requestQuery: String? = null,
        userAgent: String? = null,
        actionType: AuditActionType,
        resourceType: String? = null,
        resourceId: String? = null,
        eventContent: String,
        contextData: String? = null,
        success: Boolean
    ) {
        runCatching {
            persist(
                actorUuid = actorUuid,
                actorName = actorName,
                actorIp = actorIp,
                requestMethod = requestMethod,
                requestPath = requestPath,
                requestQuery = requestQuery,
                userAgent = userAgent,
                actionType = actionType,
                resourceType = resourceType,
                resourceId = resourceId,
                eventContent = eventContent,
                contextData = contextData,
                success = success
            )
        }
            .recoverCatching { ex ->
                val fallbackType = actionType.fallbackType() ?: throw ex
                val fallbackContent = "[fallbackFrom=$actionType] $eventContent"
                log.warn(
                    "Audit log type {} rejected by database constraint, retrying with fallback type {}",
                    actionType,
                    fallbackType
                )
                persist(
                    actorUuid = actorUuid,
                    actorName = actorName,
                    actorIp = actorIp,
                    requestMethod = requestMethod,
                    requestPath = requestPath,
                    requestQuery = requestQuery,
                    userAgent = userAgent,
                    actionType = fallbackType,
                    resourceType = resourceType,
                    resourceId = resourceId,
                    eventContent = fallbackContent,
                    contextData = contextData,
                    success = success
                )
            }
            .onFailure {
                log.error("Failed to persist audit log, type={}, actorName={}, actorIp={}", actionType, actorName, actorIp, it)
            }
    }

    @Transactional(readOnly = true)
    fun getAuditLogs(
        pageable: Pageable,
        actorName: String?,
        actionType: AuditActionType?,
        success: Boolean?,
        ip: String?,
        eventTimeFrom: LocalDateTime?,
        eventTimeTo: LocalDateTime?
    ): Page<AuditLogResponse> {
        val specification = Specification<AuditLog> { root, _, cb ->
            val predicates = mutableListOf<jakarta.persistence.criteria.Predicate>()

            actorName?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(cb.lower(root.get("actorName")), "%${keyword.lowercase()}%")
            }
            actionType?.let {
                predicates += cb.equal(root.get<AuditActionType>("actionType"), it)
            }
            success?.let {
                predicates += cb.equal(root.get<Boolean>("success"), it)
            }
            ip?.trim()?.takeIf { it.isNotEmpty() }?.let { keyword ->
                predicates += cb.like(root.get("actorIp"), "%${keyword}%")
            }
            eventTimeFrom?.let {
                predicates += cb.greaterThanOrEqualTo(root.get("eventTime"), it)
            }
            eventTimeTo?.let {
                predicates += cb.lessThanOrEqualTo(root.get("eventTime"), it)
            }

            cb.and(*predicates.toTypedArray())
        }

        return auditLogRepository.findAll(specification, pageable).map { it.toResponse() }
    }

    @Transactional(readOnly = true)
    fun getAuditLogById(id: Long): AuditLogResponse {
        val entity = auditLogRepository.findById(id)
            .orElseThrow { NoSuchElementException("审计日志不存在: id=$id") }
        return entity.toResponse()
    }

    private fun persist(
        actorUuid: UUID?,
        actorName: String?,
        actorIp: String?,
        requestMethod: String?,
        requestPath: String?,
        requestQuery: String?,
        userAgent: String?,
        actionType: AuditActionType,
        resourceType: String?,
        resourceId: String?,
        eventContent: String,
        contextData: String?,
        success: Boolean
    ) {
        auditLogRepository.saveAndFlush(
            AuditLog(
                eventTime = LocalDateTime.now(ZoneOffset.UTC),
                actorUuid = actorUuid,
                actorName = actorName,
                actorIp = actorIp,
                requestMethod = requestMethod,
                requestPath = requestPath,
                requestQuery = requestQuery,
                userAgent = userAgent,
                actionType = actionType,
                resourceType = resourceType,
                resourceId = resourceId,
                eventContent = eventContent,
                contextData = contextData,
                success = success
            )
        )
    }

    private fun AuditActionType.fallbackType(): AuditActionType? {
        return when (this) {
            AuditActionType.INSTRUCTION_CREATE -> AuditActionType.MESSAGE_CREATE
            AuditActionType.INSTRUCTION_REVIEW -> AuditActionType.MESSAGE_UPDATE
            AuditActionType.ROUND_PUBLISH -> AuditActionType.TIMELINE_UPDATE
            AuditActionType.ROUND_PAUSE -> AuditActionType.TIMELINE_UPDATE
            AuditActionType.ROUND_RESUME -> AuditActionType.TIMELINE_UPDATE
            AuditActionType.ROUND_SET_NEXT -> AuditActionType.TIMELINE_UPDATE
            AuditActionType.ROUND_UPDATE -> AuditActionType.TIMELINE_UPDATE
            AuditActionType.ROUND_SET_CURRENT -> AuditActionType.TIMELINE_JUMP
            AuditActionType.ROUND_AUTO_ADVANCE -> AuditActionType.TIMELINE_JUMP
            else -> null
        }
    }

    private fun AuditLog.toResponse(): AuditLogResponse {
        return AuditLogResponse(
            id = id ?: 0L,
            eventTime = eventTime,
            actorUuid = actorUuid?.toString(),
            actorName = actorName,
            actorIp = actorIp,
            requestMethod = requestMethod,
            requestPath = requestPath,
            requestQuery = requestQuery,
            userAgent = userAgent,
            actionType = actionType,
            resourceType = resourceType,
            resourceId = resourceId,
            eventContent = eventContent,
            contextData = contextData,
            success = success
        )
    }
}

File: backend-dotnet/src/AsyaMun.Api/Services/AuditLogService.cs
using AsyaMun.Api.Data;
using AsyaMun.Api.Dtos;
using AsyaMun.Api.Models;
using Microsoft.EntityFrameworkCore;

namespace AsyaMun.Api.Services;

public class AuditLogService
{
    private readonly IServiceScopeFactory _scopeFactory;
    private readonly AppDbContext _db;
    private readonly ILogger<AuditLogService> _logger;

    public AuditLogService(
        IServiceScopeFactory scopeFactory,
        AppDbContext db,
        ILogger<AuditLogService> logger)
    {
        _scopeFactory = scopeFactory;
        _db = db;
        _logger = logger;
    }

    public async Task SaveAsync(
        Guid? actorUuid,
        string? actorName,
        string? actorIp,
        string? requestMethod,
        string? requestPath,
        string? requestQuery,
        string? userAgent,
        AuditActionType actionType,
        string? resourceType,
        string? resourceId,
        string eventContent,
        string? contextData,
        bool success)
    {
        try
        {
            await PersistAsync(
                actorUuid, actorName, actorIp, requestMethod, requestPath, requestQuery,
                userAgent, actionType, resourceType, resourceId, eventContent, contextData, success);
        }
        catch (Exception ex)
        {
            var fallbackType = FallbackType(actionType);
            if (!fallbackType.HasValue)
            {
                _logger.LogError(ex,
                    "Failed to persist audit log, type={ActionType}, actorName={ActorName}, actorIp={ActorIp}",
                    actionType, actorName, actorIp);
                return;
            }

            _logger.LogWarning(
                "Audit log type {ActionType} rejected by database constraint, retrying with fallback type {Fallback}",
                actionType, fallbackType.Value);

            try
            {
                await PersistAsync(
                    actorUuid, actorName, actorIp, requestMethod, requestPath, requestQuery,
                    userAgent, fallbackType.Value, resourceType, resourceId,
                    $"[fallbackFrom={actionType}] {eventContent}", contextData, success);
            }
            catch (Exception ex2)
            {
                _logger.LogError(ex2,
                    "Failed to persist audit log after fallback, type={ActionType}, actorName={ActorName}",
                    fallbackType.Value, actorName);
            }
        }
    }

    private async Task PersistAsync(
        Guid? actorUuid,
        string? actorName,
        string? actorIp,
        string? requestMethod,
        string? requestPath,
        string? requestQuery,
        string? userAgent,
        AuditActionType actionType,
        string? resourceType,
        string? resourceId,
        string eventContent,
        string? contextData,
        bool success)
    {
        await using var scope = _scopeFactory.CreateAsyncScope();
        var context = scope.ServiceProvider.GetRequiredService<AppDbContext>();

        context.AuditLogs.Add(new AuditLog
        {
            EventTime = DateTime.SpecifyKind(DateTime.UtcNow, DateTimeKind.Unspecified),
            ActorUuid = actorUuid,
            ActorName = actorName,
            ActorIp = actorIp,
            RequestMethod = requestMethod,
            RequestPath = requestPath,
            RequestQuery = requestQuery,
            UserAgent = userAgent,
            ActionType = actionType,
            ResourceType = resourceType,
            ResourceId = resourceId,
            EventContent = eventContent,
            ContextData = contextData,
            Success = success
        });

        await context.SaveChangesAsync();
    }

    public async Task<Page<AuditLogResponse>> GetAuditLogsAsync(
        PageInput pageInput,
        string? actorName,
        AuditActionType? actionType,
        bool? success,
        string? ip,
        DateTime? eventTimeFrom,
        DateTime? eventTimeTo,
        CancellationToken ct = default)
    {
        var query = _db.AuditLogs.AsNoTracking();

        if (!string.IsNullOrWhiteSpace(actorName))
        {
            var keyword = actorName.Trim().ToLower();
            query = query.Where(l => l.ActorName != null && l.ActorName.ToLower().Contains(keyword));
        }

        if (actionType.HasValue)
        {
            query = query.Where(l => l.ActionType == actionType.Value);
        }

        if (success.HasValue)
        {
            query = query.Where(l => l.Success == success.Value);
        }

        if (!string.IsNullOrWhiteSpace(ip))
        {
            var keyword = ip.Trim();
            query = query.Where(l => l.ActorIp != null && l.ActorIp.Contains(keyword));
        }

        if (eventTimeFrom.HasValue)
        {
            query = query.Where(l => l.EventTime >= eventTimeFrom.Value);
        }

        if (eventTimeTo.HasValue)
        {
            query = query.Where(l => l.EventTime <= eventTimeTo.Value);
        }

        var page = await query.OrderByDescending(l => l.EventTime).ToPageAsync(pageInput, ct);
        return Page<AuditLogResponse>.Of(
            page.Content.Select(AuditLogResponse.From).ToList(),
            page.Number,
            page.Size,
            page.TotalElements);
    }

    public async Task<AuditLogResponse> GetAuditLogByIdAsync(long id, CancellationToken ct = default)
    {
        var entity = await _db.AuditLogs.AsNoTracking().FirstOrDefaultAsync(l => l.Id == id, ct)
            ?? throw new KeyNotFoundException($"审计日志不存在: id={id}");

        return AuditLogResponse.From(entity);
    }

    private static AuditActionType? FallbackType(AuditActionType actionType)
    {
        return actionType switch
        {
            AuditActionType.INSTRUCTION_CREATE => AuditActionType.MESSAGE_CREATE,
            AuditActionType.INSTRUCTION_REVIEW => AuditActionType.MESSAGE_UPDATE,
            AuditActionType.ROUND_PUBLISH => AuditActionType.TIMELINE_UPDATE,
            AuditActionType.ROUND_PAUSE => AuditActionType.TIMELINE_UPDATE,
            AuditActionType.ROUND_RESUME => AuditActionType.TIMELINE_UPDATE,
            AuditActionType.ROUND_SET_NEXT => AuditActionType.TIMELINE_UPDATE,
            AuditActionType.ROUND_UPDATE => AuditActionType.TIMELINE_UPDATE,
            AuditActionType.ROUND_SET_CURRENT => AuditActionType.TIMELINE_JUMP,
            AuditActionType.ROUND_AUTO_ADVANCE => AuditActionType.TIMELINE_JUMP,
            _ => null
        };
    }
}